instance/models/server.py

```diff
@@ -0,0 +1,274 @@
+        """
+        return self.filter(~Q(status='terminated'))
+
+
+class Server(ValidateModelMixin, TimeStampedModel, LoggerMixin):
+    """
+    A single server VM
+    """
+    status = models.CharField(max_length=11, default='new', choices=SERVER_STATUS_CHOICES, db_index=True)
+
+    objects = ServerQuerySet().as_manager()
+
+    class Meta:
+        abstract = True
+
+    def _set_status(self, status):
+        """
+        Update the current status variable, to be called when a status change is detected
+        """
+        self.status = status
+        self.log('info', 'Changed status for {}: {}'.format(self, self.status))
+        self.save()
+        return self.status
+
+    def sleep_until_status(self, target_status):
+        """
+        Sleep in a loop until the server reaches one of the specified status
+        """
+        self.log('info', 'Waiting for server {} to reach status {}...'.format(self, target_status))
+        while True:
+            self.update_status()
+            if isinstance(target_status, str):
+                if self.status == target_status:
+                    break
+            else: # Status list
+                if self.status in target_status:
+                    break
+            time.sleep(1)
+        return self.status
+
+    @staticmethod
+    def on_post_save(sender, instance, created, **kwargs):
+        """
+        Called when an instance is saved
+        """
+        publish_data('notification', {
+            'type': 'server_update',
+            'server_pk': instance.pk,
+        })
+
+    def update_status(self, provisioned=False, rebooting=False):
+        """
+        Check the current status and update it if it has changed
+        """
+        raise NotImplementedError
+
+
+class OpenStackServer(Server):
+    """
+    A Server VM hosted on an OpenStack cloud
+    """
+    instance = models.ForeignKey(OpenEdXInstance, related_name='server_set')
+    openstack_id = models.CharField(max_length=250, db_index=True, blank=True)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.nova = openstack.get_nova_client()
+
+    def __str__(self):
+        if self.openstack_id:
+            return self.openstack_id
+        else:
+            return 'New OpenStack Server'
+
+    @property
+    def os_server(self):
+        """
+        OpenStack nova server API endpoint
+        """
+        if not self.openstack_id:
+            assert self.status == 'new'
+            self.start()
+        return self.nova.servers.get(self.openstack_id)
+
+    @property
+    def public_ip(self):
+        """
+        Return one of the public address(es)
+        """
+        if not self.openstack_id:
+            return None
+
+        public_addr = openstack.get_server_public_address(self.os_server)
+        if not public_addr:
+            return None
+
+        return public_addr['addr']
+
+    def update_status(self, provisioned=False, rebooting=False):
```

Comment:
@e-kolpakov I would definitely be glad to look at a PR with this approach - I've never had a chance to play much with state machine, it would be a good opportunity to get educated : )